File: csharp/Domain/ValueObjects/Money.cs
// Delta: Closing the Specification Gap - Chapter 3 / Appendix B
//
// The [Obsolete(error:true)] constructors make constructing Money from a double or
// float a COMPILE-TIME ERROR - the single most common AI code-generation money bug.
//
// CURRENCY SAFETY MODEL (differs deliberately from Java/Python/TS):
//   The C# implementation enforces currency safety at COMPILE TIME via the TCurrency
//   type parameter - mixing Money<USD> and Money<EUR> simply will not compile, so no
//   runtime assertSameCurrency check is needed (or possible). The Java, Python, and
//   TypeScript implementations carry the currency as a value and check it at runtime.
//   Both are valid; this one catches the error earlier.
//
// CANONICAL MONEY POLICY (identical across all four language implementations):
//   - Calculation scale: 4 decimal places
//   - Display / storage scale: 2 decimal places
//   - Rounding: MidpointRounding.AwayFromZero (== HALF_UP for the values money uses)

using System.Collections.Generic;

namespace YourCompany.Domain.ValueObjects;

// -- Currency marker structs ---------------------------------------------------
// Add your currencies here. The type parameter prevents mixing USD and EUR at compile time.
public readonly struct USD { }
public readonly struct EUR { }
public readonly struct GBP { }
public readonly struct AUD { }

// -- Money<TCurrency> ----------------------------------------------------------
/// <summary>
/// Immutable monetary value with compile-time currency type safety.
///
/// USAGE:
///   var price = new Money&lt;USD&gt;(19.99m);    // correct (decimal literal)
///   var price = new Money&lt;USD&gt;(19.99);     // COMPILE ERROR (Obsolete error:true)
///   var price = new Money&lt;USD&gt;(19.99f);    // COMPILE ERROR (Obsolete error:true)
///
/// All arithmetic is computed at 4dp; use Round() or ToString() for 2dp display output.
/// </summary>
public sealed record Money<TCurrency>
    where TCurrency : struct
{
    private const int CalculationScale = 4;
    private const int DisplayScale = 2;
    private const MidpointRounding Rounding = MidpointRounding.AwayFromZero;

    /// <summary>Amount, always held at 4dp calculation scale.</summary>
    public decimal Amount { get; }

    public Money(decimal amount) =>
        Amount = Math.Round(amount, CalculationScale, Rounding);

    // -- Poison constructors - compile-time prevention -------------------------
    [Obsolete("NEVER construct Money<T> from double. Use a decimal literal: new Money<USD>(19.99m)", error: true)]
    public Money(double amount) : this((decimal)amount) { }

    [Obsolete("NEVER construct Money<T> from float. Use a decimal literal: new Money<USD>(19.99m)", error: true)]
    public Money(float amount) : this((decimal)amount) { }

    // -- Arithmetic (all at 4dp calculation scale) -----------------------------
    public Money<TCurrency> Add(Money<TCurrency> other) =>
        new(Amount + other.Amount);

    public Money<TCurrency> Subtract(Money<TCurrency> other) =>
        new(Amount - other.Amount);

    public Money<TCurrency> Multiply(decimal factor) =>
        new(Amount * factor);

    public Money<TCurrency> Divide(decimal divisor) =>
        divisor == 0m
            ? throw new DivideByZeroException("cannot divide Money by zero")
            : new(Amount / divisor);

    /// <summary>
    /// Split into <paramref name="parts"/> equal shares without losing or inventing a penny.
    /// Remainder cents are distributed one-per-share from the front, so the shares always
    /// sum back to the original at display scale. Fowler's Money-pattern allocation - a
    /// naive Divide() would silently drop the remainder.
    /// <code>new Money&lt;USD&gt;(10.00m).Allocate(3) =&gt; [3.34, 3.33, 3.33]</code>
    /// </summary>
    public IReadOnlyList<Money<TCurrency>> Allocate(int parts)
    {
        if (parts <= 0)
            throw new ArgumentOutOfRangeException(nameof(parts), "parts must be a positive integer");

        var totalCents = (long)decimal.Round(ForOutput() * 100m, 0, Rounding);
        var baseCents = totalCents / parts;
        var remainder = (int)(totalCents - baseCents * parts);

        var shares = new Money<TCurrency>[parts];
        for (var i = 0; i < parts; i++)
        {
            var shareCents = baseCents + (i < remainder ? 1 : 0);
            shares[i] = new Money<TCurrency>(shareCents / 100m);
        }
        return shares;
    }

    // -- Sign predicates -------------------------------------------------------
    public bool IsPositive => Amount > 0m;
    public bool IsZero     => Amount == 0m;
    public bool IsNegative => Amount < 0m;

    // -- Factories -------------------------------------------------------------
    public static Money<TCurrency> Zero => new(0m);
    public static Money<TCurrency> Of(decimal amount) => new(amount);

    // -- Output ----------------------------------------------------------------
    /// <summary>Amount rounded to display scale (2dp). Use for output and storage, not calculation.</summary>
    public decimal ForOutput() => Math.Round(Amount, DisplayScale, Rounding);

    /// <summary>2dp display string. Currency code is not shown - it lives only in the type parameter.</summary>
    public override string ToString() => ForOutput().ToString($"F{DisplayScale}");
}

File: java/common/Money.java
// Delta: Closing the Specification Gap - Chapter 3 / Appendix B
package com.yourcompany.common;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Currency;
import java.util.Objects;

/**
 * Immutable monetary value. Always use BigDecimal - never double or float for money.
 *
 * <pre>
 * // Correct
 * var price = Money.of("19.99", Currency.getInstance("USD"));
 * var total = price.add(tax);
 *
 * // NEVER - double arithmetic loses precision (19.99 is 19.990000000000001 in IEEE 754)
 * double price = 19.99;
 * </pre>
 *
 * CANONICAL MONEY POLICY (identical across all four language implementations):
 *   - Calculation scale: 4 decimal places
 *   - Display / storage scale: 2 decimal places
 *   - Rounding: HALF_UP
 *   - Currency mismatch: throws at runtime (C# prevents it at compile time via a type parameter).
 *
 * Note: getAmount() returns the value at 4dp calculation scale (e.g. 19.9900). For display,
 * storage, or API output, always use toDisplayAmount() which rounds to 2dp.
 */
public final class Money {

    private static final int          CALCULATION_SCALE = 4;
    private static final int          DISPLAY_SCALE     = 2;
    private static final RoundingMode ROUNDING          = RoundingMode.HALF_UP;

    private final BigDecimal amount;
    private final Currency   currency;

    private Money(BigDecimal amount, Currency currency) {
        this.currency = Objects.requireNonNull(currency, "currency must not be null");
        this.amount   = Objects.requireNonNull(amount, "amount must not be null")
                               .setScale(CALCULATION_SCALE, ROUNDING);
    }

    // -- Factories -----------------------------------------------------------
    /** Preferred factory. Use the string form to avoid floating-point errors. */
    public static Money of(String amount, Currency currency) {
        return new Money(new BigDecimal(amount), currency);
    }

    public static Money of(BigDecimal amount, Currency currency) {
        return new Money(amount, currency);
    }

    public static Money zero(Currency currency) {
        return new Money(BigDecimal.ZERO, currency);
    }

    // -- Arithmetic (all at 4dp calculation scale) ---------------------------
    public Money add(Money other) {
        assertSameCurrency(other);
        return new Money(amount.add(other.amount), currency);
    }

    public Money subtract(Money other) {
        assertSameCurrency(other);
        return new Money(amount.subtract(other.amount), currency);
    }

    public Money multiply(BigDecimal factor) {
        return new Money(amount.multiply(factor), currency);
    }

    public Money divide(BigDecimal divisor) {
        if (divisor.compareTo(BigDecimal.ZERO) == 0) {
            throw new ArithmeticException("cannot divide Money by zero");
        }
        return new Money(amount.divide(divisor, CALCULATION_SCALE, ROUNDING), currency);
    }

    /**
     * Split into {@code parts} equal shares without losing or inventing a penny.
     * Remainder cents are distributed one-per-share from the front, so the shares always
     * sum back to the original at display scale. Fowler's Money-pattern allocation - a
     * naive divide() would silently drop the remainder.
     *
     * <pre>
     * Money.of("10.00", USD).allocate(3)  ->  [3.34, 3.33, 3.33]   // sums to 10.00
     * </pre>
     */
    public java.util.List<Money> allocate(int parts) {
        if (parts <= 0) {
            throw new IllegalArgumentException("parts must be a positive integer");
        }
        long totalCents = toDisplayAmount().movePointRight(2).setScale(0).longValueExact();
        long baseCents = totalCents / parts;
        long remainder = totalCents - baseCents * parts;

        java.util.List<Money> shares = new java.util.ArrayList<>(parts);
        for (int i = 0; i < parts; i++) {
            long shareCents = baseCents + (i < remainder ? 1 : 0);
            shares.add(new Money(BigDecimal.valueOf(shareCents, 2), currency));
        }
        return shares;
    }

    // -- Output --------------------------------------------------------------
    /** 2dp for display, storage, and API output. Not for intermediate calculations. */
    public BigDecimal toDisplayAmount() {
        return amount.setScale(DISPLAY_SCALE, ROUNDING);
    }

    // -- Sign predicates -----------------------------------------------------
    public boolean isPositive() { return amount.compareTo(BigDecimal.ZERO) > 0; }
    public boolean isZero()     { return amount.compareTo(BigDecimal.ZERO) == 0; }
    public boolean isNegative() { return amount.compareTo(BigDecimal.ZERO) < 0; }

    // -- Accessors -----------------------------------------------------------
    /** Returns the amount at 4dp calculation scale. For output use toDisplayAmount(). */
    public BigDecimal getAmount()       { return amount; }
    public Currency   getCurrency()     { return currency; }
    public String     getCurrencyCode() { return currency.getCurrencyCode(); }

    // -- Internals -----------------------------------------------------------
    private void assertSameCurrency(Money other) {
        if (!currency.equals(other.currency)) {
            throw new IllegalArgumentException(
                "currency mismatch: " + currency.getCurrencyCode()
                    + " vs " + other.currency.getCurrencyCode());
        }
    }

    @Override
    public String toString() {
        return toDisplayAmount().toPlainString() + " " + currency.getCurrencyCode();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Money m)) return false;
        return amount.compareTo(m.amount) == 0 && currency.equals(m.currency);
    }

    @Override
    public int hashCode() {
        // Must be consistent with equals (which uses compareTo, ignoring scale).
        return Objects.hash(amount.stripTrailingZeros(), currency);
    }
}
